### ISSUE TITLE
0.9.0.post3 includes API not present in hidapi 0.9.0

### ISSUE THREAD

**Issue by jonasmalacofilho** [trezor/cython-hidapi#90]
`hid_get_input_report` isn't present in hidapi 0.9.0.  A new API should cause a bump on the major version number, and I imagine this will happen when on the next hidapi release.

Besides the issue with the version number, the use of this API also makes the current release unpackageable on many Linux distributions, due to their use of `--with-system-hidapi`.  ArchLinux, Debian and Fedora are affected, and those are probably not the only ones.

_(Note: hidapi is no longer unmaintained.)_

ArchLinux is especially impacted here since it already tried to ship this new version, leading to this:

```
$ python -c "import hid"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: /home/jonas/Code/archlinux/python-hidapi/trunk/src/hidapi-0.9.0.post3/build/lib.linux-x86_64-3.8/hid.cpython-38-x86_64-linux-gnu.so: undefined symbol: hid_get_input_report
```

Related: [ArchLinux bug FS#67103](https://bugs.archlinux.org/task/67103) ([python-hidapi] hid_get_input_report not present in hidapi (dependency mismatch))  
Related: jonasmalacofilho/liquidctl#144 ("ImportError: undefined symbol: hid_get_input_report on ArchLinux")

---

Honestly, I'm not sure what can be done here.

Ideally the build system would only implement `device.get_input_report` if `hid_get_input_report` is available or, alternatively, if `--with-system-hidapi` has _not_ been passed.

But I can't find an easy way to do conditional compilation in a Cython `pyd`... maybe I missed something in the docs.

**Comment by prusnak** [comment]
> (Note: hidapi is no longer unmaintained.)

Not true. Hidapi has found a new home here: https://github.com/libusb/hidapi

I think the best way how to proceed from here is to ask upstream to release a new version of hidapi, ideally with a bumped version to signify the API change.

**Comment by jonasmalacofilho** [comment]
Because it's now expected to have semi-regular releases I find it a lot harder to push distros to use an arbitrary commit of hidapi, either for their own hidapi package or just in python-hidapi (by removing `--with-system-hidapi`).

> Not true. Hidapi has found a new home[...]

The fact that it has a new home is what I meant by "no longer unmaintained"; the double negation made it equivalent to "maintained".

**Comment by prusnak** [comment]
> Because it's now expected to have semi-regular releases

Ok, let's wait for the next release then.

**Comment by Foxboron** [comment]
This exchange still leaves out the important question: *why* is this project suddenly depending on unreleased APIs from another project?

**Comment by prusnak** [comment]
There is no good answer to your question. Maybe the closest one to good is that the cython-hidapi is a project not funded by anyone and my priorities are elsewhere. We stopped using hidapi for Trezor quite some time ago, but I still keep maintaining the package in my free time. I asked the libusb folks whether they wont to adopt the project, but the answer was no.

If you want to help you can send a PR with a fix, I will gladly review it. But then it was you who said that hidapi will see a release soon, so I am not sure whether that's worth the effort.

**Comment by Foxboron** [comment]
>If you want to help you can send a PR with a fix, I will gladly review it.

It would entail reverting the patch until hidapi has a release. It would just be simpler to revert the patch in Arch completely for the time being.

>But then it was you who said that hidapi will see a release soon, so I am not sure whether that's worth the effort.

You are confusing me with the original author of the issue, I'm the package maintainer for Arch.

**Comment by prusnak** [comment]
> It would entail reverting the patch until hidapi has a release. It would just be simpler to revert the patch in Arch completely for the time being.

Ok, please do. Why did you (or whoever) bump the version of cython-hidapi in Arch if it did not build? In my book, this is a responsibility of a packager to not break their distro in case upstream screws up.

> You are confusing me with the original author of the issue, I'm the package maintainer for Arch.

Ok, but the point still stands.

**Comment by Foxboron** [comment]
>Why did you (or whoever) bump the version of cython-hidapi in Arch if it did not build? 

It builds fine with just a warning.

```
==> Starting build()...
running build
running build_ext
cythoning hidraw.pyx to hidraw.c
building 'hid' extension
creating build
creating build/temp.linux-x86_64-3.8
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/hidapi -I/usr/include/python3.8 -c hidraw.c -o build/temp.linux-x86_64-3.8/hidraw.o
hidraw.c: In function ‘__pyx_pf_3hid_6device_24get_input_report’:
hidraw.c:4631:21: warning: implicit declaration of function ‘hid_get_input_report’; did you mean ‘hid_get_feature_report’? [-Wimplicit-function-declaration]
 4631 |         __pyx_v_n = hid_get_input_report(__pyx_v_c_hid, __pyx_v_cbuff, __pyx_v_c_max_length);
      |                     ^~~~~~~~~~~~~~~~~~~~
      |                     hid_get_feature_report
creating build/lib.linux-x86_64-3.8
gcc -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fno-semantic-interposition -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.8/hidraw.o -L/usr/lib -ludev -lrt -lhidapi-hidraw -o build/lib.linux-x86_64-3.8/hid.cpython-38-x86_64-linux-gnu.so
/usr/lib/python3.8/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /build/python-hidapi/src/hidapi-0.9.0.post3/hidraw.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)
```

Which is somewhat unexpected when there is a complete API missing.

However, if we run the one test the project provides, the issue is obvious with an error.

```
======================================================================
ERROR: tests (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: tests
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.8/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/build/python-hidapi/src/hidapi-0.9.0.post3/tests.py", line 3, in <module>
    import hid
ImportError: /build/python-hidapi/src/hidapi-0.9.0.post3/hid.cpython-38-x86_64-linux-gnu.so: undefined symbol: hid_get_input_report
```

>In my book, this is a responsibility of a packager to not break their distro in case upstream screws up.

Sure, but it's unexpected to have upstream suddenly depend on unreleased APIs :upside_down_face:

**Comment by prusnak** [comment]
@jonasmalacofilho @hegjon @Foxboron You can update packages in Fedora/Arch/whatever to just released hidapi-0.10.0 and cython-hidapi-0.10.0.

**Comment by hegjon** [comment]
> @jonasmalacofilho @hegjon @Foxboron You can update packages in Fedora/Arch/whatever to just released hidapi-0.10.0 and cython-hidapi-0.10.0.

Thanks for the heads-up

**Comment by hegjon** [comment]
Updated and built for Fedora https://koji.fedoraproject.org/koji/taskinfo?taskID=54422819
